plaidml/edsl/tests/edsl_test.cc

```diff
@@ -1957,6 +1957,137 @@ TEST_F(CppEdsl, Lens) {
   checkExact(program, {input}, {expected});
 }
 
+TEST_F(CppEdsl, Loop) {
+  auto A = Placeholder(DType::FLOAT32, {4});
+  int64_t lb = 0;
+  int64_t hb = 5;
+  int64_t step = 1;
+  Tensor O = loop(/*lowerBound*/ lb, /*upperBound*/ hb, /*step*/ step,  //
+                  /*loop-carried variable*/ {A},                        //
+                  [&]() { return A + 1; });
+  auto program = makeProgram("loop", {A}, {O});
+  std::vector<float> input = {
+      1, 1, 1, 1  //
+  };
+  std::vector<float> expected = {
+      6, 6, 6, 6  //
+  };
+  checkExact(program, {input}, {expected});
+}
+
+TEST_F(CppEdsl, LoopConstantBuffer) {
+  auto A = Placeholder(DType::FLOAT32, {4});
+  size_t lb = 0;
+  size_t hb = 5;
+  size_t step = 1;
+  Tensor O = loop(/*lowerBound*/ lb, /*upperBound*/ hb, /*step*/ step,  //
+                  /*loop-carried variable*/ {A},                        //
+                  [&]() {
+                    std::vector<float> test{1, 1, 1, 1};
+                    auto B = Constant(makeBuffer(DType::FLOAT32, {4}, test), "test");
+                    return A + B;
+                  });
+  auto program = makeProgram("loop", {A}, {O});
+  std::vector<float> input = {
+      1, 1, 1, 1  //
+  };
+  std::vector<float> expected = {
+      6, 6, 6, 6  //
+  };
+  checkExact(program, {input}, {expected});
+}
+
+TEST_F(CppEdsl, LoopMultiIter) {
+  auto A = Placeholder(DType::FLOAT32, {4});
+  auto B = Placeholder(DType::FLOAT32, {4});
+  int64_t lb = 0;
+  int64_t hb = 5;
+  int64_t step = 1;
+  auto fn = [&]() -> TensorVec {
+    auto O = A + 1;
+    auto T = B + 1;
+    return {O, T};
+  };
+  auto output = loop(/*lowerBound*/ lb, /*upperBound*/ hb, /*step*/ step,  //
+                     /*loop-carried variable*/ {A, B}, /*loop inner function*/ fn);
+  auto O = output[0] + output[1];
+  auto program = makeProgram("loop", {A, B}, {O});
+  std::vector<float> input1 = {
+      1, 1, 1, 1  //
+  };
+  std::vector<float> input2 = {
+      0, 0, 0, 0  //
+  };
+  std::vector<float> expected1 = {
+      11, 11, 11, 11  //
+  };
+  checkExact(program, {input1, input2}, {expected1});
+}
+
+TEST_F(CppEdsl, LoopWithBeforeOp) {
+  auto A = Placeholder(DType::FLOAT32, {4});
+  auto B = Placeholder(DType::FLOAT32, {4});
+  auto C = A + 1;
+  auto D = C * B;
```

Comment:
the code like` A = A + 1 ` will cause the `missing placeholder` issues while making program, and it also arise in other test cases. so I use `auto C = A + 1` to replace it. I thought it can represent the same mean.

Reply:
I see, thanks. Could you add buffer checking in these new tests?

Reply:
ok, fixed this.